File: go/walletconnect/caip122.go
// Package walletconnect is the Go port of @luxwallet/connect's login verifier.
//
// It mirrors the TypeScript src/ byte-for-byte so Hanzo IAM (Casdoor, Go)
// verifies wallet logins identically to verifyProof: same CAIP-122 message
// parse, same domain/nonce/time checks, same EIP-191 digest, same ed25519
// check. The two implementations must stay in lockstep.
package walletconnect

import (
	"fmt"
	"regexp"
	"strings"
)

// Chain is a supported chain family. Values, not places — namespaced by this
// type. Mirrors the TS `Chain` union.
type Chain string

const (
	ChainEVM     Chain = "evm"
	ChainSolana  Chain = "solana"
	ChainBitcoin Chain = "bitcoin"
	ChainTON     Chain = "ton"
	ChainXRP     Chain = "xrp"
)

// chainLabel is the human chain label used on the first line of the message.
// Mirrors the TS CHAIN_LABEL record exactly.
var chainLabel = map[Chain]string{
	ChainEVM:     "Ethereum",
	ChainSolana:  "Solana",
	ChainBitcoin: "Bitcoin",
	ChainTON:     "TON",
	ChainXRP:     "XRP Ledger",
}

// ParsedSiwx is the result of parsing a CAIP-122 message. Optional fields use
// pointers so absence (TS `undefined`) is distinguishable from empty string.
type ParsedSiwx struct {
	Domain         string
	Address        string
	Statement      *string
	URI            string
	Version        *string
	ChainID        *string
	Nonce          string
	IssuedAt       string
	ExpirationTime *string
	NotBefore      *string
	RequestID      *string
	Resources      []string
}

// LoginChallenge mirrors the TS LoginChallenge: the fields a server asks a
// wallet to sign. Optional fields are pointers (TS `undefined`).
type LoginChallenge struct {
	Domain         string
	URI            string
	Statement      *string
	Nonce          string
	IssuedAt       string
	ExpirationTime *string
	NotBefore      *string
	RequestID      *string
	Version        *string
	Resources      []string
}

// BuildParams mirrors the TS BuildParams for buildSiwxMessage.
type BuildParams struct {
	Challenge LoginChallenge
	Address   string
	Chain     Chain
	// ChainID is the CAIP-2 network id, e.g. "eip155:1". nil to omit.
	ChainID *string
}

// BuildSiwxMessage renders a LoginChallenge to the canonical CAIP-122 message
// string. Mirrors the TS buildSiwxMessage line-for-line. Returns an error if a
// statement contains a newline (TS throws).
func BuildSiwxMessage(p BuildParams) (string, error) {
	c := p.Challenge
	label := chainLabel[p.Chain]

	lines := make([]string, 0, 16)
	lines = append(lines, fmt.Sprintf("%s wants you to sign in with your %s account:", c.Domain, label))
	lines = append(lines, p.Address)
	lines = append(lines, "")
	// Statement block is optional. When present it sits on its own line between
	// two blank lines (per EIP-4361 ABNF).
	if c.Statement != nil && len(*c.Statement) > 0 {
		if strings.Contains(*c.Statement, "\n") {
			return "", fmt.Errorf("caip122: statement must be a single line")
		}
		lines = append(lines, *c.Statement)
		lines = append(lines, "")
	}
	lines = append(lines, "URI: "+c.URI)
	version := "1"
	if c.Version != nil {
		version = *c.Version
	}
	lines = append(lines, "Version: "+version)
	if p.ChainID != nil {
		lines = append(lines, "Chain ID: "+*p.ChainID)
	}
	lines = append(lines, "Nonce: "+c.Nonce)
	lines = append(lines, "Issued At: "+c.IssuedAt)
	if c.ExpirationTime != nil {
		lines = append(lines, "Expiration Time: "+*c.ExpirationTime)
	}
	if c.NotBefore != nil {
		lines = append(lines, "Not Before: "+*c.NotBefore)
	}
	if c.RequestID != nil {
		lines = append(lines, "Request ID: "+*c.RequestID)
	}
	if len(c.Resources) > 0 {
		lines = append(lines, "Resources:")
		for _, r := range c.Resources {
			lines = append(lines, "- "+r)
		}
	}
	return strings.Join(lines, "\n"), nil
}

// headerRE mirrors the TS HEADER_RE. The domain is non-greedy (`+?`) so it
// stops at the first " wants you to sign in with your " occurrence. Go's RE2
// has no lazy/greedy backtracking difference that matters here because the
// suffix is anchored with `$`, but we keep `+?` for fidelity with the TS source.
var headerRE = regexp.MustCompile(`^(.+?) wants you to sign in with your .+ account:$`)

// fieldRE mirrors the TS FIELD_RE: an exact, ordered set of recognised keys.
var fieldRE = regexp.MustCompile(`^(URI|Version|Chain ID|Nonce|Issued At|Expiration Time|Not Before|Request ID): (.*)$`)

func strPtr(s string) *string { return &s }

// ParseSiwxMessage parses a CAIP-122 message back into its fields. Returns an
// error on malformed input. Mirrors the TS parseSiwxMessage exactly, including
// the statement/resources block handling and the required-field check.
func ParseSiwxMessage(message string) (ParsedSiwx, error) {
	var out ParsedSiwx
	raw := strings.Split(message, "\n")
	if len(raw) < 2 {
		return out, fmt.Errorf("caip122: message too short")
	}
	header := headerRE.FindStringSubmatch(raw[0])
	if header == nil {
		return out, fmt.Errorf("caip122: malformed header line")
	}
	out.Domain = header[1]
	out.Address = strings.TrimSpace(raw[1])
	if len(out.Address) == 0 {
		return out, fmt.Errorf("caip122: missing address line")
	}

	// Everything from line 2 onward: an optional statement block, then fields.
	var resources []string
	inResources := false
	var statementParts []string
	sawField := false

	for i := 2; i < len(raw); i++ {
		line := raw[i]
		if inResources {
			if strings.HasPrefix(line, "- ") {
				resources = append(resources, line[2:])
				continue
			}
			inResources = false
		}
		if line == "Resources:" {
			inResources = true
			sawField = true
			continue
		}
		if f := fieldRE.FindStringSubmatch(line); f != nil {
			sawField = true
			key, v := f[1], f[2]
			switch key {
			case "URI":
				out.URI = v
			case "Version":
				out.Version = strPtr(v)
			case "Chain ID":
				out.ChainID = strPtr(v)
			case "Nonce":
				out.Nonce = v
			case "Issued At":
				out.IssuedAt = v
			case "Expiration Time":
				out.ExpirationTime = strPtr(v)
			case "Not Before":
				out.NotBefore = strPtr(v)
			case "Request ID":
				out.RequestID = strPtr(v)
			}
			continue
		}
		// Pre-field, non-empty, non-field lines are the statement.
		if !sawField && len(line) > 0 {
			statementParts = append(statementParts, line)
		}
	}

	if len(statementParts) > 0 {
		out.Statement = strPtr(strings.Join(statementParts, "\n"))
	}
	if len(resources) > 0 {
		out.Resources = resources
	}

	// TS checks: out.uri == null || out.nonce == null || out.issuedAt == null.
	// A field that never appeared is the zero value "" here; the TS message
	// always renders these three, and parse must reject a message missing any.
	if out.URI == "" || out.Nonce == "" || out.IssuedAt == "" {
		return out, fmt.Errorf("caip122: missing required field (URI / Nonce / Issued At)")
	}
	return out, nil
}

File: src/caip122.ts
/**
 * CAIP-122 "Sign-In-With-X" message — the one canonical login string for every
 * chain. Generalizes EIP-4361 (SIWE) so a Solana / Bitcoin / TON / XRP wallet
 * signs the exact same human-readable assertion an Ethereum wallet does.
 *
 * Spec: https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-122.md
 * (which itself generalizes https://eips.ethereum.org/EIPS/eip-4361)
 *
 * build ∘ parse round-trips. Both are pure — no I/O, no clock — so the Go port
 * for IAM can mirror them byte-for-byte.
 */
import type { Chain, LoginChallenge } from './types.js';

/** Human chain label used on the first line of the message. */
const CHAIN_LABEL: Record<Chain, string> = {
  evm: 'Ethereum',
  solana: 'Solana',
  bitcoin: 'Bitcoin',
  ton: 'TON',
  xrp: 'XRP Ledger',
  polkadot: 'Polkadot',
  cardano: 'Cardano',
};

export interface ParsedSiwx {
  domain: string;
  address: string;
  statement?: string;
  uri: string;
  version?: string;
  chainId?: string;
  nonce: string;
  issuedAt: string;
  expirationTime?: string;
  notBefore?: string;
  requestId?: string;
  resources?: string[];
}

export interface BuildParams {
  challenge: LoginChallenge;
  address: string;
  chain: Chain;
  /** CAIP-2 network id, e.g. 'eip155:1', 'solana:5eykt...'. Optional. */
  chainId?: string;
}

/** Render a {@link LoginChallenge} to the canonical CAIP-122 message string. */
export function buildSiwxMessage(params: BuildParams): string {
  const { challenge: c, address, chain, chainId } = params;
  const label = CHAIN_LABEL[chain];

  const lines: string[] = [];
  lines.push(`${c.domain} wants you to sign in with your ${label} account:`);
  lines.push(address);
  lines.push('');
  // Statement block is optional. When present it sits on its own line between
  // two blank lines (per EIP-4361 ABNF).
  if (c.statement != null && c.statement.length > 0) {
    if (c.statement.includes('\n')) {
      throw new Error('caip122: statement must be a single line');
    }
    lines.push(c.statement);
    lines.push('');
  }
  lines.push(`URI: ${c.uri}`);
  lines.push(`Version: ${c.version ?? '1'}`);
  if (chainId != null) {
    lines.push(`Chain ID: ${chainId}`);
  }
  lines.push(`Nonce: ${c.nonce}`);
  lines.push(`Issued At: ${c.issuedAt}`);
  if (c.expirationTime != null) {
    lines.push(`Expiration Time: ${c.expirationTime}`);
  }
  if (c.notBefore != null) {
    lines.push(`Not Before: ${c.notBefore}`);
  }
  if (c.requestId != null) {
    lines.push(`Request ID: ${c.requestId}`);
  }
  if (c.resources != null && c.resources.length > 0) {
    lines.push('Resources:');
    for (const r of c.resources) {
      lines.push(`- ${r}`);
    }
  }
  return lines.join('\n');
}

const HEADER_RE = /^(?<domain>[^\n]+?) wants you to sign in with your .+ account:$/;
const FIELD_RE = /^(?<key>URI|Version|Chain ID|Nonce|Issued At|Expiration Time|Not Before|Request ID): (?<val>.*)$/;

/** Parse a CAIP-122 message back into its fields. Throws on malformed input. */
export function parseSiwxMessage(message: string): ParsedSiwx {
  const raw = message.split('\n');
  if (raw.length < 2) {
    throw new Error('caip122: message too short');
  }
  const header = HEADER_RE.exec(raw[0] ?? '');
  if (!header?.groups) {
    throw new Error('caip122: malformed header line');
  }
  const domain = header.groups.domain;
  const address = (raw[1] ?? '').trim();
  if (address.length === 0) {
    throw new Error('caip122: missing address line');
  }

  // Everything from line 2 onward: an optional statement block, then fields.
  const out: Partial<ParsedSiwx> = { domain, address };
  const resources: string[] = [];
  let inResources = false;
  let statementParts: string[] = [];
  let sawField = false;

  for (let i = 2; i < raw.length; i++) {
    const line = raw[i] ?? '';
    if (inResources) {
      if (line.startsWith('- ')) {
        resources.push(line.slice(2));
        continue;
      }
      inResources = false;
    }
    if (line === 'Resources:') {
      inResources = true;
      sawField = true;
      continue;
    }
    const f = FIELD_RE.exec(line);
    if (f?.groups) {
      sawField = true;
      const v = f.groups.val;
      switch (f.groups.key) {
        case 'URI': out.uri = v; break;
        case 'Version': out.version = v; break;
        case 'Chain ID': out.chainId = v; break;
        case 'Nonce': out.nonce = v; break;
        case 'Issued At': out.issuedAt = v; break;
        case 'Expiration Time': out.expirationTime = v; break;
        case 'Not Before': out.notBefore = v; break;
        case 'Request ID': out.requestId = v; break;
      }
      continue;
    }
    // Pre-field, non-empty, non-field lines are the statement.
    if (!sawField && line.length > 0) {
      statementParts.push(line);
    }
  }

  if (statementParts.length > 0) {
    out.statement = statementParts.join('\n');
  }
  if (resources.length > 0) {
    out.resources = resources;
  }

  if (out.uri == null || out.nonce == null || out.issuedAt == null) {
    throw new Error('caip122: missing required field (URI / Nonce / Issued At)');
  }
  return out as ParsedSiwx;
}
